exec/check-lockfiles-up-to-date/src/checkLockfilesUpToDate.ts

```diff
@@ -0,0 +1,387 @@
+
+  if (!virtualStoreDir) return
+
+  const rootManifestOptions = rootProjectManifest && rootProjectManifestDir
+    ? getOptionsFromRootManifest(rootProjectManifestDir, rootProjectManifest)
+    : undefined
+
+  if (allProjects && workspaceDir) {
+    const packagesList = await loadPackagesList({ cacheDir, workspaceDir })
+    if (!packagesList) {
+      throw new PnpmError('RUN_CHECK_DEPS_NO_CACHE', 'Cannot check whether dependencies are outdated', {
+        hint: 'Run `pnpm install` to create the cache',
+      })
+    }
+
+    if (!equals(
+      filter(value => !!value, packagesList.catalogs ?? {}),
+      filter(value => !!value, catalogs ?? {})
+    )) {
+      throw new PnpmError('RUN_CHECK_DEPS_OUTDATED', 'Catalogs cache outdated', {
+        hint: 'Run `pnpm install` to update the catalogs cache',
+      })
+    }
+
+    const currentProjectRootDirs = allProjects.map(project => project.rootDir).sort()
+    if (!equals(packagesList.projectRootDirs, currentProjectRootDirs)) {
+      throw new PnpmError('RUN_CHECK_DEPS_WORKSPACE_STRUCTURE_CHANGED', 'The workspace structure has changed since last install', {
+        hint: 'Run `pnpm install` to update the workspace structure and dependencies tree',
+      })
+    }
+
+    const allManifestStats = await Promise.all(allProjects.map(async project => {
+      const manifestStats = await statManifestFile(project.rootDir)
+      if (!manifestStats) {
+        // this error should not happen
+        throw new Error(`Cannot find one of ${MANIFEST_BASE_NAMES.join(', ')} in ${project.rootDir}`)
+      }
+      return { project, manifestStats }
+    }))
+
+    const modifiedProjects = allManifestStats.filter(
+      ({ manifestStats }) =>
+        manifestStats.mtime.valueOf() > packagesList.lastValidatedTimestamp
+    )
+
+    if (modifiedProjects.length === 0) {
+      logger.debug({ msg: 'No manifest files are modified after the last validation. Exiting check.' })
+      return
+    }
+
+    logger.debug({ msg: 'Some manifest files are modified after the last validation. Continuing check.' })
+
+    let readWantedLockfileAndDir: (projectDir: string) => Promise<{
+      wantedLockfile: Lockfile
+      wantedLockfileDir: string
+    }>
+    if (sharedWorkspaceLockfile) {
+      const wantedLockfileStats = await statIfExists(path.join(workspaceDir, WANTED_LOCKFILE))
+      if (!wantedLockfileStats) return throwLockfileNotFound(workspaceDir)
+
+      const wantedLockfilePromise = readWantedLockfile(workspaceDir, { ignoreIncompatible: false })
+      if (wantedLockfileStats.mtime.valueOf() > packagesList.lastValidatedTimestamp) {
+        const virtualStoreDir = path.join(workspaceDir, 'node_modules', '.pnpm')
+        const currentLockfile = await readCurrentLockfile(virtualStoreDir, { ignoreIncompatible: false })
+        const wantedLockfile = (await wantedLockfilePromise) ?? throwLockfileNotFound(workspaceDir)
+        assertLockfilesEqual(currentLockfile, wantedLockfile, workspaceDir)
+      }
```

Comment:
To optimize it even more we can create a function that reads and compares both lockfiles. It could first just compare the sizes of the lockfile files, without reading and parsing them. If the sizes don’t match we know that the node_modules is not up to date. If the sizes match we can compare the raw contents, no need to parse the files (unless we want to print info about how exactly the don’t match).

Reply:
Should we increase the complexity of the code to optimize for an error case?